### PR TITLE
Use a shadowDom for the Vomnibar (and all UIComponents)

### PR DESCRIPTION
**Pull request by mrmr1993**
This wraps `UIComponent.iframeElement` in a shadow DOM, so that page CSS doesn't override our own.

@smblott-github suggest this instead of #1613?

### CODE REVIEW

**Comment by smblott-github** [comment]
Thanks, @mrmr1993.  I'm not sure if you're finished yet (a new commit popped up since I last looked). However, at 1d16e4e996aa5234020a73212d74834ba99ec162, this isn't actually working for http://www.motorbeam.com/.

(Nothing interesting in the console.)

**Comment by mrmr1993** [comment]
Sorry, I think I've found the error (a typo), I'm testing now and then I'll push.

**Comment by mrmr1993** [comment]
@smblott-github I've updated this, it should be working as intended now

File: content_scripts/ui_component.coffee [hunk 1]
```diff
@@ -5,12 +5,29 @@ class UIComponent
...
   constructor: (iframeUrl, className, @handleMessage) ->
+    styleSheet = document.createElement "style"
+    styleSheet.type = "text/css"
+    # Default to everything hidden while the stylesheet loads.
+    styleSheet.innerHTML = "* {display: none !important;}"
+    # Load stylesheet.
+    xhr = new XMLHttpRequest()
+    xhr.onload = (e) -> styleSheet.innerHTML = xhr.responseText
+    xhr.open "GET", chrome.runtime.getURL("content_scripts/vimium.css"), true
+    xhr.send()
```
Comment (smblott-github) [inline content_scripts/ui_component.coffee]: Why do you need the `XMLHttpRequest` for the style sheet but not for the iframe contents?

(Looks odd.)

Comment (mrmr1993) [inline content_scripts/ui_component.coffee]: Shadow DOM doesn't support `<link>` elements, but it does support `<iframes>`.

**Comment by smblott-github** [comment]
> Shadow DOM doesn't support <link> elements

Well it does seem to, at least to some extent.  One of [these two lines](https://github.com/philc/vimium/blob/master/pages/vomnibar.html#L11) is needed (and one apparently not).

**Comment by mrmr1993** [comment]
> Well it does seem to, at least to some extent.

You can't have a `<link>` element in the shadow DOM (at least, not in Chrome). The iframe DOM is just a normal DOM and can contain all the usual things, though.

> One of [these two lines](https://github.com/philc/vimium/blob/master/pages/vomnibar.html#L11) is needed (and one apparently not).

Don't we need styles for `.vimiumReset` from `vimium.css` to keep the base style consistent with the rest of the extension? Also, the contents of the iframe should be orthogonal to this PR...?

**Comment by smblott-github** [comment]
@mrmr1993.  This seems to work nicely.  I'm not seeing any issues.
We have three options for 1.51 (which will probably happen quite soon -- bug fixes):

1. Stick with #1613 (it works for the original use case).
1. Go with this.

You think we should go with this *and now*, @mrmr1993?

Also... with the Vomnibar now effectively a web component (yes?), what benefit are we getting from involving an iframe?

**Comment by mrmr1993** [comment]
> You think we should go with this *and now*, @mrmr1993?

Yes.

> what benefit are we getting from involving an iframe?

* It retains the current focused element and selection (#599, which can be closed), so bookmarklets can use them.
* Having an in-page Vomnibar lets the page open arbitrary URLs, extract bookmarks, etc. by manipulating the `<input>` and its contents.
* Dealing with `<input>`s in shadow DOMs is fiddly (#853).

**Pull request merged**